File: morale-assistant-backend/src/main/java/hu/herolds/projects/moraleassistant/service/SynthesizerService.java
package hu.herolds.projects.moraleassistant.service;

import hu.herolds.projects.moraleassistant.exception.SynthesizeException;
import hu.herolds.projects.moraleassistant.model.enums.Language;
import hu.herolds.projects.moraleassistant.util.synthesizer.Synthesizer;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections.CollectionUtils;
import org.springframework.stereotype.Service;

import java.nio.file.Path;
import java.util.*;

@Slf4j
@Service
public class SynthesizerService {
    private Map<Language, Set<Synthesizer>> synthesizerServiceMap;

    public void initialize(final List<Synthesizer> synthesizers) {
        synthesizerServiceMap = new HashMap<>();

        synthesizers.forEach(synthesizerService -> synthesizerService.getSupportedLanguages().forEach(language -> {
            if (!synthesizerServiceMap.containsKey(language)) {
                synthesizerServiceMap.put(language, new HashSet<>());
            }
            synthesizerServiceMap.get(language).add(synthesizerService);
            log.info("Registered synthesizer service to language {}: {}", language, synthesizerService.getClass().getSimpleName());
        }));
    }

    public Path synthesize(final Language language, final String text) throws SynthesizeException {
        final Synthesizer synthesizer = getSynthesizerService(language);
        return synthesizer.synthesize(text);
    }

    private Synthesizer getSynthesizerService(final Language language) throws SynthesizeException {
        final Set<Synthesizer> synthesizers = synthesizerServiceMap.get(language);

        if (CollectionUtils.isEmpty(synthesizers)) {
            throw new SynthesizeException("Language is not supported: " + language);
        }

        return new ArrayList<>(synthesizers).get(0);
    }
}

File: morale-assistant-backend-kt/src/main/kotlin/hu/herolds/projects/morale/service/SynthesizerService.kt
package hu.herolds.projects.morale.service

import hu.herolds.projects.morale.domain.enums.Language
import hu.herolds.projects.morale.exception.SynthesizeException
import hu.herolds.projects.morale.service.synthetize.Synthesizer
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class SynthesizerService {
    private val log = LoggerFactory.getLogger(javaClass)

    private var synthesizerMap: MutableMap<Language, MutableSet<Synthesizer>> = mutableMapOf()

    fun initialize(synthesizers: List<Synthesizer>) {
        synthesizers.forEach {synthesizer: Synthesizer ->
            synthesizer.supportedLanguages.forEach { language ->
                synthesizerMap.computeIfAbsent(language) {
                    mutableSetOf()
                }

                synthesizerMap[language]!!.add(synthesizer)
                log.info("Registered synthesizer service to language $language: ${synthesizer.javaClass.simpleName}")
            }
        }
    }

    fun synthesize(language: Language, text: String): ByteArray = getSynthesizerService(language).synthesize(text)

    private fun getSynthesizerService(language: Language): Synthesizer = synthesizerMap[language]?.first()
            ?: throw SynthesizeException("Language is not supported: $language")
}